### PR TITLE
feat(sheets-thread-comment): comment support for sheets

### PR DESCRIPTION
**Pull request by weird94**
close #xxx, #yyy, #zzzz

### CODE REVIEW

**Comment by weird94** [comment]
https://github.com/dream-num/univer/pull/2228#issuecomment-2106574167

File: examples/src/sheets/main.ts [hunk 1]
```diff
@@ -100,6 +101,31 @@ if (!IS_E2E) {
...
+const mockUser = {
+    userID: 'mockId',
+    name: 'MockUser',
+    avatar: 'data:image/png;base64,iVBORw0KGgoAAAANSUhEUgAAABgAAAAYCAYAAADgdz34AAAACXBIWXMAAAsTAAALEwEAmpwYAAAAAXNSR0IArs4c6QAAAARnQU1BAACxjwv8YQUAAAInSURBVHgBtZU9TxtBEIbfWRzFSIdkikhBSqRQkJqkCKTCFkqVInSUSaT0wC8w/gXxD4gU2nRJkXQWhAZowDUUWKIwEgWWbEEB3mVmx3dn4DA2nB/ppNuPeWd29mMIPXDr+RxwtgRHeW6+guNPRxogqnL7Dwz9psJ27S4NShaeZTH3kwXy6I81dlRKcmRui88swdq9AcSFL7Buz1Vmlns64MiLsCjzwnIYHLH57tbfFbs7KRaXyEU8FVZofqccOfA5l7Q8LPIkGrwnb2RPNEXWFVMUF3L+kDCk0btDDAMzOm5YfAHDwp4tG74wnzAsiOYMnJ3GoDybA7IT98/jm5+JNnfiIzAS6LlqHQBN/i6b2t/cV1Hh6BfwYlHnHP4AXi5q/8kmMMpOs8+BixZw/Fd6xUEHEbnkgclvQP2fGp7uShRKnQ3G32rkjV1th8JhIGG7tR/JyjGteSOZELwGMmNqIIigRCLRh2OZIE6BjItdd7pCW6Uhm1zzkUtungSxwEUzNpQ+GQumtH1ej1MqgmNT6vwmhCq5yuwq56EYTbgeQUz3yvrpV1b4ok3nYJ+eYhgYmjRUqErx2EDq0Fr8FhG++iqVGqxlUJI/70Ar0UgJaWHj6hYVHJrfKssAHot1JfqwE9WVWzXZVd5z2Ws/4PnmtEjkXeKJDvxUecLbWOXH/DP6QQ4J72NS0adedp1aseBfXP8odlZFfPvBF7SN/8hky1TYuPOAXAEipMx15u5ToAAAAABJRU5ErkJggg==',
+};
+
+univer.registerPlugin(UniverSheetsThreadCommentPlugin, {
+    mentions: [{
+        trigger: '@',
+        mentions: [
```
Comment (wzhudev) [inline examples/src/sheets/main.ts:113]: Are you intent to inject mock data here? I suggest remove this API in avoid of confusion to other developers.

File: packages/design/src/components/button/Button.tsx [hunk 2]
```diff
@@ -60,6 +60,8 @@ export interface IButtonProps {
 
     /** Set the handler to handle `click` event */
     onClick?: (e: React.MouseEvent<HTMLButtonElement>) => void;
+
+    id?: string;
```
Comment (wzhudev) [inline packages/design/src/components/button/Button.tsx:64]: Should extend native button attributes on this interface. @jikkai

File: packages/core/src/services/user-manager/user-manager.service.ts [hunk 3]
```diff
@@ -15,14 +15,28 @@
  */
 
 import type { IUser } from '@univerjs/protocol';
-import { Subject } from 'rxjs';
+import { BehaviorSubject, Subject } from 'rxjs';
 
 
 export class UserManagerService {
```
Comment (wzhudev) [inline packages/core/src/services/user-manager/user-manager.service.ts:21]: Maybe this service should be moved out of the core pacakge.

File: packages/sheets-thread-comment/src/commands/operations/comment.operation.ts [hunk 4]
```diff
@@ -0,0 +1,71 @@
...
+        const threadCommentPanelService = accessor.get(ThreadCommentPanelService);
+        const activeCell = selectionManagerService.getFirst()?.primary;
+        const current = selectionManagerService.getCurrent();
+        const model = accessor.get(SheetsThreadCommentModel);
+        if (!current || !activeCell) {
+            return false;
+        }
+
+        const { unitId, sheetId } = current;
+        const workbook = univerInstanceService.getUnit<Workbook>(current.unitId, UniverInstanceType.UNIVER_SHEET);
```
Comment (wzhudev) [inline packages/sheets-thread-comment/src/commands/operations/comment.operation.ts:41]: Use `getSheetCommandTarget` API to simplify the code.

Comment (wzhudev) [reply packages/sheets-thread-comment/src/commands/operations/comment.operation.ts:41]: Can this work with live share? Have you tested that?

File: packages/sheets-thread-comment/src/controllers/menu.ts [hunk 5]
```diff
@@ -0,0 +1,52 @@
...
+import type { IMenuItem, IShortcutItem } from '@univerjs/ui';
+import { getMenuHiddenObservable, KeyCode, MenuItemType, MenuPosition, MetaKeys } from '@univerjs/ui';
+import type { IAccessor } from '@wendellhu/redi';
+import { ToggleSheetCommentPanelOperation } from '@univerjs/thread-comment-ui';
+import { UniverInstanceType } from '@univerjs/core';
+import { whenSheetEditorFocused } from '@univerjs/sheets-ui';
+import { ShowAddSheetCommentModalOperation } from '../commands/operations/comment.operation';
+import { COMMENT_SINGLE_ICON } from '../types/const';
+
+export const threadCommentMenu = (accessor: IAccessor) => {
```
Comment (wzhudev) [inline packages/sheets-thread-comment/src/controllers/menu.ts:26]: The names of menu factories should adhere to our naming conversion. Refer to menu factories in `sheets` package.

File: packages/sheets-thread-comment/src/controllers/sheets-thread-comment.controller.ts [hunk 6]
```diff
@@ -0,0 +1,162 @@
...
+        this._initCommandListener();
+        this._initPanelListener();
+    }
+
+    private _initShortcut() {
+        this._shortcutService.registerShortcut(AddCommentShortcut);
+    }
+
+    private _initCommandListener() {
+        this._commandService.onCommandExecuted((commandInfo) => {
```
Comment (wzhudev) [inline packages/sheets-thread-comment/src/controllers/sheets-thread-comment.controller.ts:63]: `this._disposeWithMe`

File: packages/sheets-thread-comment/src/controllers/sheets-thread-comment.controller.ts [hunk 7]
```diff
@@ -0,0 +1,162 @@
...
+            }
+        });
+    }
+
+    private _initMenu() {
+        [
+            threadCommentMenu,
+            threadPanelMenu,
+        ].forEach((menu) => {
+            this._menuService.addMenuItem(menu(this._injector));
```
Comment (wzhudev) [inline packages/sheets-thread-comment/src/controllers/sheets-thread-comment.controller.ts:95]: Same as above. Please pay attention to disposing logics.

File: packages/sheets-thread-comment/src/controllers/sheets-thread-comment.controller.ts [hunk 8]
```diff
@@ -0,0 +1,162 @@
...
+        });
+    }
+
+    private _initComponent() {
+        ([
+            [SHEETS_THREAD_COMMENT_MODAL, SheetsThreadCommentCell],
+            [THREAD_COMMENT_PANEL, SheetsThreadCommentPanel],
+            [COMMENT_SINGLE_ICON, CommentSingle],
+        ] as const).forEach(([key, comp]) => {
+            this._componentManager.register(key, comp);
```
Comment (wzhudev) [inline packages/sheets-thread-comment/src/controllers/sheets-thread-comment.controller.ts:105]: Same here.

File: packages/thread-comment/src/controllers/tc-remove-subunit.controller.ts [hunk 9]
```diff
@@ -0,0 +1,77 @@
...
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+import type { Workbook } from '@univerjs/core';
+import { Disposable, IUniverInstanceService, LifecycleStages, OnLifecycle, UniverInstanceType } from '@univerjs/core';
+import { Inject } from '@wendellhu/redi';
+import type { IRemoveSheetCommandParams } from '@univerjs/sheets';
+import { RemoveSheetCommand, SheetInterceptorService } from '@univerjs/sheets';
```
Comment (wzhudev) [inline packages/thread-comment/src/controllers/tc-remove-subunit.controller.ts:21]: If this package is for general purposes, it shoud not rely on `@univerjs/sheets`.

File: packages/thread-comment/src/types/interfaces/i-thread-comment.ts [hunk 10]
```diff
@@ -0,0 +1,47 @@
...
+
+export interface IThreadCommentMention {
+    type: string;
+    label: string;
+    id: string;
+    icon?: string;
+    extra?: any;
+}
+
+export type TextNode = {
```
Comment (wzhudev) [inline packages/thread-comment/src/types/interfaces/i-thread-comment.ts:25]: Is is compatible with our own rich text format? In the future we may should migrate the editor to our own. cc @Jocs

File: packages/sheets-thread-comment/src/controllers/sheets-thread-comment-render.controller.ts [hunk 11]
```diff
@@ -0,0 +1,89 @@
...
+import { Disposable, IUniverInstanceService, LifecycleStages, OnLifecycle, UniverInstanceType } from '@univerjs/core';
+import { INTERCEPTOR_POINT, SheetInterceptorService } from '@univerjs/sheets';
+import { Inject } from '@wendellhu/redi';
+import { SheetSkeletonManagerService } from '@univerjs/sheets-ui';
+import type { Spreadsheet } from '@univerjs/engine-render';
+import { IRenderManagerService } from '@univerjs/engine-render';
+import { SheetsThreadCommentModel } from '../models/sheets-thread-comment.model';
+
+@OnLifecycle(LifecycleStages.Ready, SheetsThreadCommentRenderController)
+export class SheetsThreadCommentRenderController extends Disposable {
```
Comment (wzhudev) [inline packages/sheets-thread-comment/src/controllers/sheets-thread-comment-render.controller.ts:27]: This should be refactored to `RenderController`. Please refer to exsiting implementations in `packages/sheets-ui/src/controllers/render-controllers`.

**Review by wzhudev — CHANGES_REQUESTED** [review]
Some problems need to be addressed in avoid of breaking changes in the future.

File: packages/thread-comment/src/types/interfaces/i-thread-comment.ts (hunk 10 above)
Comment (Jocs) [reply packages/thread-comment/src/types/interfaces/i-thread-comment.ts:25]: Need to convert to our rich text format, why not just use our rich text format?

File: packages/thread-comment/src/controllers/tc-remove-subunit.controller.ts (hunk 9 above)
Comment (weird94) [reply packages/thread-comment/src/controllers/tc-remove-subunit.controller.ts:21]: moved to sheets-thread-comment

**Comment by weird94** [comment]
https://github.com/dream-num/univer/pull/2228#issuecomment-2106574167

**Review by wzhudev — APPROVED** [review]
LGTM

**Pull request merged**